utils/gridlink_impl.c.src

```diff
@@ -179,113 +142,158 @@ cellarray_DOUBLE * gridlink_DOUBLE(const int64_t NPART,
         if (ix>nmesh_x-1)  ix--;    /* this shouldn't happen, but . . . */
         if (iy>nmesh_y-1)  iy--;
         if (iz>nmesh_z-1)  iz--;
-        XRETURN(X[i] >= xmin && X[i] <= xmax, NULL,
-               "x[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-                i, X[i], xmin, xmax);
-        XRETURN(Y[i] >= ymin && Y[i] <= ymax, NULL,
-               "y[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-               i, Y[i], ymin, ymax);
-        XRETURN(Z[i] >= zmin && Z[i] <= zmax, NULL,
-               "z[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-               i, Z[i], zmin, zmax);
-
-        XRETURN(ix >= 0 && ix < nmesh_x, NULL, "ix=%d must be within [0,%d)\n", ix, nmesh_x);
-        XRETURN(iy >= 0 && iy < nmesh_y, NULL, "iy=%d must be within [0,%d)\n", iy, nmesh_y);
-        XRETURN(iz >= 0 && iz < nmesh_z, NULL, "iz=%d must be within [0,%d)\n", iz, nmesh_z);
+    
+        out_of_bounds += ix < 0 || ix >= nmesh_x || iy < 0 || iy >= nmesh_y || iz < 0 || iz >= nmesh_z;
 
         const int64_t icell = CONVERT_3D_INDEX_TO_LINEAR(ix, iy, iz, nmesh_x, nmesh_y, nmesh_z);
-        if(options->copy_particles == 0) {
-            all_cell_indices[i] = icell;
+        all_cell_indices[i] = icell;
+        if(!options->copy_particles){
             original_indices[i] = i;
+        }
+    }
+    
+    XRETURN(out_of_bounds == 0, NULL, "%"PRId64" particles are out of bounds. Check periodic wrapping?\n", out_of_bounds);
+    
+    // Now determine the number of particles per cell so each cell knows its global starting index
+    // This information is used for both the copy and in-place versions
+#if defined(_OPENMP)
+    int Nthread = omp_get_max_threads();
+#else
+    int Nthread = 1;
+#endif
+    int64_t *cell_occupation[Nthread];
+
+#if defined(_OPENMP)
+    #pragma omp parallel for schedule(static)
+#endif
+    for(int t = 0; t < Nthread; t++){
+        cell_occupation[t] = my_calloc(totncells, sizeof(int64_t));
+    }
+#if defined(_OPENMP)
+    #pragma omp parallel for schedule(static)
+#endif
+    for(int64_t i = 0; i < NPART; i++){
+        #if defined(_OPENMP)
+        int tid = omp_get_thread_num();
+        #else
+        int tid = 0;
+        #endif
+        cell_occupation[tid][all_cell_indices[i]]++;
```

Comment:
This could result in false-sharing and cache flushes

Reply:
`cell_occupation` is an array of pointers, so each thread has its own allocation to work in (one malloc per thread, line 170).  In fact, avoiding false sharing is exactly why it's structured this way!

Technically, a few elements could share a cache line across threads at the very end/beginning of the arrays if the mallocs aren't suitably padded.  But even if they aren't padded, it must be a minuscule effect for so few elements...

Reply:
Right - I see! But the way the malloc is implemented, the allocation could be done by a different thread. Think it will be better to declare an OMP region and hoist the `tid` at the top of that OMP region, then a single call to malloc `cell_occupation[tid]`, and then start the OMP loop over `NPART`

Reply:
The mallocs are done in an OpenMP loop above to ensure this doesn't happen.  The loop is statically scheduled and has a number of iterations equal to the number of threads, so every thread gets one iteration (guaranteed by the spec).  Technically, chunk_size of 1 should be specified, too, to be super safe about it.  But I can see that putting it all in one parallel region might be clearer, so I'll go ahead and change that.